proposals/lip-0041.md

```diff
@@ -97,23 +103,216 @@ In this section, we describe the properties of an auth account and specify their
+def verify(trs: Transaction) -> None:
+    # Check that a multisignature group is not already registered for this address
+    senderAddress = 20-byte address derived from trs.senderPublicKey
+    if len(authAccount(senderAddress).mandatoryKeys) > 0 or len(authAccount(senderAddress).mandatoryKeys) > 0:
+        raise Exception('Multisignature group already registered for this address.')
+    # Mandatory keys are pairwise distinct and ordered lexicographically.
+    if trs.params.mandatoryKeys != sorted(set(trs.params.mandatoryKeys)):
+        raise Exception('Invalid mandatory keys array.')
+    # Optional keys are pairwise distinct and ordered lexicographically.
+    if trs.params.optionalKeys != sorted(set(trs.params.optionalKeys)):
+        raise Exception('Invalid optional keys array.')
+    # Mandatory and optional keys must be disjoint.
+    all_keys = trs.params.mandatoryKeys + trs.params.optionalKeys
+    if len(all_keys) != len(set(all_keys)):
+        raise Exception('Mandatory and optional keys arrays are not disjoint.')
+    # Union of mandatory and optional keys must contain between 1 and 64 keys (included).
+    if len(all_keys) < 1 or len(all_keys) > 64:
+        raise Exception('Invalid length of mandatory and optional keys arrays.')
+    # Number of signatures must be an integer between 1 and the total number of public keys.
+    if trs.params.numberOfSignatures < 1 or trs.params.numberOfSignatures > len(all_keys):
+        raise Exception('Invalid number of signatures.')
+    # Number of signatures must not be smaller than the number of mandatory keys.
+    if trs.params.numberOfSignatures < len(trs.params.mandatoryKeys):
+        raise Exception('Invalid number of signatures.')
+
+    # Verify the signatures 
+    message = encode(
+        multisigRegMsgSchema,
+        {
+            "address": senderAddress,
+            "nonce": trs.nonce,
+            "numberOfSignatures": trs.params.numberOfSignatures,
+            "mandatoryKeys": trs.params.mandatoryKeys,
+            "optionalKeys": trs.params.optionalKeys
+        }
+    )
+    for signature, key in zip(signatures, all_keys):
+        verifyMessageSig(key, MULTISIGNATURE_MSG_TAG, networkIdentifier, message, signature)
+```
+
+The function `verifyMessageSig` is defined in [LIP 0062](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0062.md#specification) and `networkIdentifier` is the network identifier of the chain.
+ schema `multisignatureGroupDataSchema` is 
 
 ##### Execution
 
-This command updates the entries in the sender `authAccount` to the values specified in the command parameters.
+```python
+def execute(trs: Transaction) -> None:
+    senderAddress = 20-byte address derived from trs.senderPublicKey
+    authAccount(senderAddress).numberOfSignatures = trs.params.numberOfSignatures
+    authAccount(senderAddress).mandatoryKeys = trs.params.mandatoryKeys
+    authAccount(senderAddress).optionalKeys = trs.params.optionalKeys
+    
+    emitEvent(
```

Comment:
What is the rationale for emitting this event if it only contains data already included in the transaction and the only topic is the transaction sender?

Reply:
I think it would be nice if the state changes are communicated though events

Reply:
agree (I think we solved this on slack)